lite2/client.go

```diff
@@ -0,0 +1,374 @@
+		nextVals      *types.ValidatorSet
+		err           error
+	)
+	for height := c.trustedHeader.Height + 1; height < newHeader.Height; height++ {
+		interimHeader, err = c.primary.SignedHeader(height)
+		if err != nil {
+			return errors.Wrapf(err, "failed to obtain the header #%d", height)
+		}
+
+		err = Verify(c.chainID, c.trustedHeader, c.trustedNextVals, interimHeader, c.trustedNextVals, c.trustingPeriod, now, c.trustLevel)
+		if err != nil {
+			return errors.Wrapf(err, "failed to verify the header #%d", height)
+		}
+
+		// Update trusted header and vals.
+		if height == newHeader.Height-1 {
+			nextVals = newVals
+		} else {
+			nextVals, err = c.primary.ValidatorSet(height + 1)
+			if err != nil {
+				return errors.Wrapf(err, "failed to obtain the vals #%d", height+1)
+			}
+		}
+		if !bytes.Equal(interimHeader.NextValidatorsHash, nextVals.Hash()) {
+			return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
+				interimHeader.NextValidatorsHash,
+				nextVals.Hash(),
+				height)
+		}
+		err = c.updateTrustedHeaderAndVals(interimHeader, nextVals)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update trusted state #%d", height)
+		}
+	}
+
+	// 2) Verify the new header.
+	return Verify(c.chainID, c.trustedHeader, c.trustedNextVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+}
+
+func (c *Client) bisection(
+	lastHeader *types.SignedHeader,
+	lastVals *types.ValidatorSet,
+	newHeader *types.SignedHeader,
+	newVals *types.ValidatorSet,
+	now time.Time) error {
+
+	err := Verify(c.chainID, lastHeader, lastVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+	switch err.(type) {
+	case nil:
+		return nil
+	case types.ErrTooMuchChange:
+		// continue bisection
+	default:
+		return errors.Wrapf(err, "failed to verify the header #%d ", newHeader.Height)
+	}
+
+	if newHeader.Height == c.trustedHeader.Height+1 {
+		// TODO: submit evidence here
+		return errors.Errorf("adjacent headers (#%d and #%d) that are not matching", lastHeader.Height, newHeader.Height)
+	}
+
+	pivot := (c.trustedHeader.Height + newHeader.Header.Height) / 2
+	pivotHeader, pivotVals, err := c.fetchHeaderAndValsAtHeight(pivot)
+	if err != nil {
+		return err
+	}
+
+	if err := c.bisection(lastHeader, lastVals, pivotHeader, pivotVals, now); err == nil {
+		nextVals, err := c.primary.ValidatorSet(pivot + 1)
+		if err != nil {
+			return errors.Wrapf(err, "failed to obtain the vals #%d", pivot+1)
+		}
+		if !bytes.Equal(pivotHeader.NextValidatorsHash, nextVals.Hash()) {
+			return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
+				pivotHeader.NextValidatorsHash,
+				nextVals.Hash(),
+				pivot)
+		}
+		return c.bisection(pivotHeader, nextVals, newHeader, newVals, now)
+	}
+
+	return errors.New("bisection failed. Restart with different full-node?")
+}
+
+func (c *Client) updateTrustedHeaderAndVals(h *types.SignedHeader, vals *types.ValidatorSet) error {
```

Comment:
IMO this function should verify the validator set against the header's `NextValSetHash`, unless there is a reason we would ever not want to do that.